### ISSUE TITLE
Make JavaScript work in Java 8

### ISSUE THREAD

**Issue by ctrueden** [imagej/imagej#116]
Java 8 removed the [Rhino](http://mozilla.org/rhino/) implementation of JavaScript in favor of the newer, faster, better, shinier, and backwards-incompatible [Nashorn](http://openjdk.java.net/projects/nashorn/) implementation.

So most (probably all) JavaScript scripts distributed with ImageJ and Fiji do not work in Java 8.

Let's fix that by updating the SciJava [scripting-javascript](https://github.com/scijava/scripting-javascript) component to bundle the [Rhino implementation](http://search.maven.org/#artifactdetails%7Corg.mozilla%7Crhino%7C1.7R5%7Cjar) again, and use it by default. We will likely have to pillage the [JSR-223-compliant JavaScript script engine implementation](http://fiji.sc/git/?p=scripting.git;a=tree;f=engines/javascript/src/com/sun/phobos/script/javascript;h=a66a32e15fd3fb1ddff8004829922b88077d19c4;hb=refs/heads/svn/trunk) from the [java.net Scripting](http://fiji.sc/git/?p=scripting.git;a=summary) project, too. And pray that it works the same as the one bundled with Java 6 and Java 7...

Furthermore, fixing JavaScript in ImageJ2 (i.e., the SciJava scripting framework, used by the Script Editor, etc.) is all well and good. But what about ImageJ1? It is highly likely that it calls JavaScript in a different way which will also not work with Java 8, and that we will thus have to do more runtime patching to fix things to work properly again with certain legacy workflows.

Related questions:
* Should there be separate "JavaScript (Rhino)" and "JavaScript (Nashorn)" entries in the Script Editor?
* If so, what the extension should we use for Nashorn scripts (maybe `.nashorn.js`, with Nashorn having a higher priority than the Rhino-based script language?).

**Comment by ctrueden** [comment]
Fortunately, we didn't have to resort to shipping our own Rhino interpreter, because Nashorn provides a reasonable [backwards compatibility layer](http://www.oracle.com/technetwork/articles/java/jf14-nashorn-2126515.html). Using that seems to do the trick, at least on the surface.

**Comment by ctrueden** [comment]
Note that as things stand, when running ImageJ with Java 8 on OS X, the following error is given:
```
[ERROR] Cannot create plugin: class='org.scijava.plugins.scripting.javascript.JavaScriptScriptLanguage', name='JavaScript', priority=0.0, enabled=true, pluginType=ScriptLanguage
java.lang.IllegalArgumentException: No such script engine: javascript
	at org.scijava.script.AdaptedScriptLanguage.findFactory(AdaptedScriptLanguage.java:151)
	at org.scijava.script.AdaptedScriptLanguage.<init>(AdaptedScriptLanguage.java:73)
	at org.scijava.plugins.scripting.javascript.JavaScriptScriptLanguage.<init>(JavaScriptScriptLanguage.java:59)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at java.lang.Class.newInstance(Class.java:442)
	at org.scijava.plugin.PluginInfo.createInstance(PluginInfo.java:302)
	at org.scijava.plugin.DefaultPluginService.createInstance(DefaultPluginService.java:236)
	at org.scijava.plugin.DefaultPluginService.createInstances(DefaultPluginService.java:225)
	at org.scijava.plugin.DefaultPluginService.createInstancesOfType(DefaultPluginService.java:216)
	at org.scijava.plugin.AbstractSingletonService.initInstances(AbstractSingletonService.java:123)
	at org.scijava.plugin.AbstractSingletonService.getInstances(AbstractSingletonService.java:72)
	at org.scijava.script.DefaultScriptService.initScriptLanguageIndex(DefaultScriptService.java:342)
	at org.scijava.script.DefaultScriptService.scriptLanguageIndex(DefaultScriptService.java:307)
	at org.scijava.script.DefaultScriptService.getIndex(DefaultScriptService.java:117)
	at org.scijava.script.DefaultScriptService.canHandleFile(DefaultScriptService.java:233)
	at org.scijava.script.ScriptFinder.discoverScripts(ScriptFinder.java:131)
	at org.scijava.script.ScriptFinder.discoverScripts(ScriptFinder.java:125)
	at org.scijava.script.ScriptFinder.findScripts(ScriptFinder.java:93)
	at org.scijava.script.DefaultScriptService.initScripts(DefaultScriptService.java:391)
	at org.scijava.script.DefaultScriptService.scripts(DefaultScriptService.java:325)
	at org.scijava.script.DefaultScriptService.getScripts(DefaultScriptService.java:167)
	at net.imagej.legacy.LegacyService.getScriptsAndNonLegacyCommands(LegacyService.java:627)
	at net.imagej.legacy.IJ1Helper.addMenuItems(IJ1Helper.java:630)
	at net.imagej.legacy.LegacyService.initialize(LegacyService.java:501)
	at org.scijava.service.ServiceHelper.createServiceRecursively(ServiceHelper.java:353)
	at org.scijava.service.ServiceHelper.createExactService(ServiceHelper.java:269)
	at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:231)
	at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:194)
	at org.scijava.service.ServiceHelper.loadServices(ServiceHelper.java:166)
	at org.scijava.Context.<init>(Context.java:267)
	at org.scijava.Context.<init>(Context.java:224)
	at org.scijava.Context.<init>(Context.java:163)
	at org.scijava.Context.<init>(Context.java:119)
	at org.scijava.Context.<init>(Context.java:107)
	at net.imagej.ImageJ.<init>(ImageJ.java:82)
	at net.imagej.Main.launch(Main.java:61)
	at net.imagej.Main.main(Main.java:73)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at net.imagej.launcher.ClassLauncher.launch(ClassLauncher.java:258)
	at net.imagej.launcher.ClassLauncher.run(ClassLauncher.java:184)
	at net.imagej.launcher.ClassLauncher.main(ClassLauncher.java:76)
```

This error is caused by the fact that the ImageJ launcher passes `-Djava.ext.dirs=/Applications/Fiji.app/java/macosx-java3d/Home/lib/ext:/Library/Java/Extensions:/System/Library/Java/Extensions:/System/Library/Frameworks/JavaVM.framework/Home/lib/ext` to the Java process. To fix it, we'll either need to fix the launcher code, or migrate to a different launch mechanism—probably JavaFX's self-contained application bundles.

**Comment by ctrueden** [comment]
(For completeness, the above "No such script engine" error was fixed in imagej/imagej-launcher@4e1e688906d140c3ea6313ca2a0f9cc3b5879644 and imagej/imagej-launcher@333ade43b38dfa3295641c9e6f70801808b16480.)
